cryptographic_estimators/MQEstimator/series/hilbert.py

```diff
@@ -102,90 +135,106 @@ def precision(self):
-            sage: H.series
-            1 + 4*z + z^2 - 16*z^3 - 14*z^4 + 40*z^5 + 50*z^6 - 80*z^7 - 125*z^8 + 140*z^9 + O(z^10)
+            sage: H.coefficient_of_degree(5)
+            -4
+
         """
-        return self._series
 
-    @property
-    def series_up_to_degree(self):
+        if d < self._prec:
+            return int(self._series[d])
+
+        raise ValueError(
+            f"The degree d should be smaller than the precision of the series which is {self._prec}"
+        )
+
+    def coefficient_up_to_degree(self, d: int):
         """
-        Return the series `self.series(z)/(1-z)`
+        Return the d-th coefficient in the Hilbert serie `self._series/(1-x)`
 
         EXAMPLES::
 
             sage: from cryptographic_estimators.MQEstimator.series.hilbert import HilbertSeries
             sage: H = HilbertSeries(4, [2]*5)
-            sage: H.series_up_to_degree
-            1 + 5*z + 10*z^2 + 10*z^3 + 5*z^4 + z^5 + O(z^10)
+            sage: H.coefficient_up_to_degree(4)
+            5
+
         """
-        z = self._ring.gen()
-        return self._series/(1-z)
 
-    @property
-    def npolynomials(self):
+        if d < self._prec:
+            return int(self._series_up_to_degree[d])
+        raise ValueError(
+            f"The degree d should be smaller than the precision of the series which is {self._prec}"
+        )
+
+    def first_nonpositive_coefficient(self):
         """
-        Return the no. of polynomials
+        Return the first non-positive integer of the series
 
         EXAMPLES::
 
             sage: from cryptographic_estimators.MQEstimator.series.hilbert import HilbertSeries
             sage: H = HilbertSeries(10, [2]*15)
-            sage: H.npolynomials
-            15
+            sage: H.first_nonpositive_coefficient()
+            4
+
         """
-        return len(self._degrees)
+        serie = self._series
+        for d in range(self.precision):
+            if serie[d] <= 0:
+                return int(d)
+        raise ValueError("Unable to find a nonpositive coefficient in the serie.")
 
-    def first_nonpositive_integer(self):
+    def first_nonpositive_coefficient_up_to_degree(self):
         """
-        Return the first non-positive integer of the series
+        Return the first non-positive integer of the serie `self._series/(1-x)`
+
 
         EXAMPLES::
 
             sage: from cryptographic_estimators.MQEstimator.series.hilbert import HilbertSeries
             sage: H = HilbertSeries(10, [2]*15)
-            sage: H.first_nonpositive_integer()
-            4
+            sage: H.first_nonpositive_coefficient_up_to_degree()
+            5
+
         """
-        s = self.series()
         for d in range(self.precision):
-            if s[d] <= 0:
-                return ZZ(d)
+            if self._series_up_to_degree[d] <= 0:
+                return int(d)
         raise ValueError(
-            "unable to find a nonpositive coefficient in the series")
```

Comment:
is it good to directly raise an exception? Im not familiar with Hilbert series, is it common that a series  does not contain a nonpositive coefficient? Or is this a rather rare thing to happen?
If its common, I think its a bad Idea to directly raise an exception. Maybe one could return 0?

Reply:
BTW: is it a good idea to raise exception in the "hot" code at all?

Reply:
You are raising a good point. I'm afraid returning 0 might lead a misinterpretations of the results of the estimator. I think if we set the "Error" case to 0, in order to avoid misinterpretations, we have to check if the output of this method is > than 0 every time we call it, and I think this is worse than raising an error. 

When calling that function the user is expecting a nonzero degree. I think it is better make it aware of if when such a degree cannot be returned.

Reply:
We don't expect an MQEstimator user to always to use `try catch`.

Reply:
ok than its fine be me